WIP - Pallet revive benchmarks

lrubasze opened the pull request:
## pallet-revive: Add and refine `finalize_block` benchmarks

This PR introduces proper benchmarking support for `pallet-revive`’s `finalize_block` logic and performs cleanup, fixes, and adjustments to enable reliable weight measurement.

### Key changes
- **Benchmarking support**
  - Added initial `finalize_block` benchmarks.
  - Introduced `on_finalize_internal` / `on_initialize_internal` to allow direct benchmark invocation of finalization logic.
  - Provided valid signed transaction setup in benchmarks to simulate realistic execution.
  - Adjusted `block_author()` to return a valid author account when `runtime-benchmarks` feature is enabled.

- **Runtime / crypto adjustments**
  - Temporarily enabled `sp-application-crypto`’s `full_crypto` feature to support benchmarking key usage.

### Impact
This lays the groundwork for applying accurate weights to `finalize_block` by:
- Measuring the cost of its operations in a controlled environment.
- Enabling future tuning for both constant and per-transaction costs.
- Ensuring benchmarks use valid transactions and block authors for realistic measurement.

lrubasze commented on substrate/frame/revive/src/benchmarking.rs:
```diff
@@ -2296,6 +2302,104 @@ mod benchmarks {
+		signed_tx.signed_payload()
+	}
+
+	// `c`: number of transactions to fit in the block
```
Added benchmarks to measure weight per event and data size.

pgherveou commented on substrate/frame/revive/src/benchmarking.rs:
```diff
@@ -2330,6 +2336,283 @@ mod benchmarks {
+	}
+
+	// Macro to generate common finalize_block benchmark setup
+	macro_rules! setup_finalize_block_benchmark {
```
looks like you could use a regular function instead of a macro can't you?

lrubasze replied:
Since there is yet another benchmark `finalize_block_event_data_processing` I am leaving it as a separate function. If you decide to merge both benchmarks into one I will inline it.

pgherveou commented on substrate/frame/revive/src/benchmarking.rs:
```diff
@@ -2330,6 +2336,283 @@ mod benchmarks {
+
+	// `c`: number of transactions to fit in the block
+	#[benchmark(pov_mode = Measured)]
+	fn finalize_block_transaction_processing(c: Linear<0, 100>) -> Result<(), BenchmarkError> {
```
at first glance I am wondering why you have 2 benchmark with a single parameter, rather than one with the two paramters

this would generate a weight function similar to  
finalize_block_processing(events, transaction)

Then you can deduct:
- base cost          finalize_block_processing(0, 0)
- cost of 1 event  finalize_block_processing(1, 0) - finalize_block_processing(0, 0)
- cost of 1 tx        finalize_block_processing(0, 1) - finalize_block_processing(0, 0)

lrubasze replied:
Merged `finalize_block_transaction_processing(transaction_cnt)` and `finalize_block_event_processing(event_cnt)` into single  `finalize_block_processing(transaction_cnt, event_cnt)`.

Please keep in mind that there is yet another benchmark `finalize_block_event_data_processing(data_len)`. I left it as a separate one, for the sake of simplicity. But if you want I can merge it with the upper one as well.

lrubasze commented on substrate/frame/revive/src/benchmarking.rs:
```diff
@@ -2330,6 +2336,283 @@ mod benchmarks {
+			current_block,
+		) = setup_finalize_block_benchmark!();
+
+		#[block]
```
Yes, that makes sense.
But there are still some operations in eth_call which supposedly shall be benchmarked. I mean reading `InflightEvents` and writing `InflightTransactions` in `store_transaction` method. 
Or can we just skip them?

lrubasze replied:
Hm, maybe it should be something like this
```rust
	// Initialize block
	let _ = Pallet::<T>::on_initialize(current_block);

	#[block]
	{
            // Store executed transactions
		    let _ = Pallet::<T>::store_transaction(signed_payload.clone(), true, gas_consumed);

			// Finalize block
            let _ = Pallet::<T>::on_finalize(current_block);

		}
```

lrubasze replied:
Done

lrubasze commented on substrate/frame/revive/src/benchmarking.rs:
```diff
@@ -2330,6 +2336,283 @@ mod benchmarks {
+		let unsigned_tx: TransactionUnsigned = TransactionLegacyUnsigned {
+			to: Some(target_address),
+			value,
+			chain_id: Some(1.into()),
```
done

pgherveou commented on substrate/frame/revive/src/lib.rs:
```diff
@@ -1762,6 +1822,11 @@ impl<T: Config> Pallet<T> {
 	/// Therefore all events must be contract emitted events.
 	fn deposit_event(event: Event<T>) {
 		if matches!(event, Event::ContractEmitted { .. }) {
+			frame_system::Pallet::<T>::register_extra_weight_unchecked(
```
yeah that's not the right way of doing this, we need to charge our contract gas meter 
fortunately we have already a way to charge for events
you can just add the extra weight here

https://github.com/paritytech/polkadot-sdk/blob/lrubasze/pallet-revive-benchmarks/substrate/frame/revive/src/vm/runtime_costs.rs?plain=1#L258 and leave this untouched

I think this code added below as well with InflightEvents should probably be moved to

https://github.com/paritytech/polkadot-sdk/blob/lrubasze/pallet-revive-benchmarks/substrate/frame/revive/src/exec.rs?plain=1#L2018-L2024

cc @lexnv

lrubasze commented:
PR slightly refactor after discussion with @pgherveou.

Introduced benchmarks:
- `on_finalize(n, p)` - Measures the cost of adding transactions to a block during finalization
- `on_finalize_per_transaction(e)` - Measures the cost of processing events within transaction during finalization

Additionally transaction payload size is taken into account when measuring the cost of transaction.

lexnv commented on substrate/frame/revive/src/benchmarking.rs:
```diff
@@ -2330,6 +2336,245 @@ mod benchmarks {
+		}
+
+		// Verify block was properly finalized
+		assert!(Pallet::<T>::block_author().is_some());
```
Same question as above

lexnv commented on substrate/frame/revive/src/weights_utils.rs:
```diff
@@ -0,0 +1,76 @@
+	///
+	/// This uses differential cost calculation to isolate the marginal cost components:
+	/// per_event_base + (data_len * per_byte_cost)
+	fn on_finalize_block_per_event(data_len: u32) -> Weight {
```
nit: Is this taking into account both `event.data` and `event.topics`? In other words, is this the sum of the two?

lrubasze commented on substrate/frame/revive/src/lib.rs:
```diff
@@ -1542,12 +1567,16 @@ where
 
+		// Safe to unwrap(), it is asserted when converting to TransactionUnsigned above
 		let eth_transact_call =
-			crate::Call::<T>::eth_transact { payload: unsigned_tx.dummy_signed_payload() };
```
Ok, I have rebased but I will leave like this and not the default transaction.
Real tx is needed here in order to determine costs properly.

pgherveou commented on substrate/frame/revive/src/benchmarking.rs:
```diff
@@ -2330,6 +2336,245 @@ mod benchmarks {
+	/// - Per event: `on_finalize_per_transaction(1, 0) - on_finalize_per_transaction(0, 0)`
+	/// - Per byte of event data: `on_finalize_per_transaction(1, d) -
+	///   on_finalize_per_transaction(1, 0)`
+	#[benchmark(pov_mode = Measured)]
+	fn on_finalize_per_transaction(
+		e: Linear<0, 100>,
+		d: Linear<0, 16384>,
+	) -> Result<(), BenchmarkError> {
```
I wonder if that work with the benchmark framework, or if we should create another benchmark as you initially suggested.

@ggwpez is this correct to create a benchmark with two variable here where `e` is the number of events and `d` the total bytes of data for distributed over all the events.

rather than doing one benchmark with `e` events and a fixed data and another one with 1 event and `d` bytes of data

lrubasze replied:
Done

pgherveou commented on substrate/frame/revive/src/benchmarking.rs:
```diff
@@ -2330,6 +2336,245 @@ mod benchmarks {
+				let event_data_size = bytes_per_event + if i == 0 { extra_bytes } else { 0 };
+				let event_data = vec![0x42u8; event_data_size as usize];
+
+				Pallet::<T>::deposit_event(Event::<T>::ContractEmitted {
```
It’s more accurate to benchmark the storage write directly here as @lexnv suggested,
since that’s the data we actually want to measure. 

Using `Pallet::<T>::deposit_event` adds indirection
it could change in the future to only emit a Substrate event without performing the storage write

lrubasze replied:
Ok, to clarify. The part which is in fact benchmarked here is the `#[block]` section in [line 2559](https://github.com/paritytech/polkadot-sdk/pull/9496/files#diff-fcd7ecb9d9eaf4542b7613d3428f39faf65b97753d53d08face51b89fb27b734R2559):

```rust
	#[block]
	{
		// Measure the finalization cost with specific event/data parameters
		let _ = Pallet::<T>::on_finalize(current_block);
	}
```

`deposit_event()` (or `store_transaction()` mentioned above) are only used to prepare for benchmark. 
@lexnv @pgherveou do you still think it needs some change?

lrubasze commented on substrate/frame/revive/src/benchmarking.rs:
```diff
@@ -2330,6 +2336,245 @@ mod benchmarks {
+			.into_iter()
+			.filter(|event_record| {
+				// Use string representation to identify contract events
+				format!("{:?}", event_record.event).contains("ContractEmitted")
```
I was thinking of having some basic sanity check after benchmark completes. But probably you're right that's not needed. I will remove it

lrubasze commented:
Ok, I've addressed some comments.
I am waiting with rebasing until we get some answers related this [comment](https://github.com/paritytech/polkadot-sdk/pull/9418#discussion_r2294024816).
cc: @lexnv @pgherveou

lrubasze commented on substrate/frame/revive/src/benchmarking.rs:
```diff
@@ -2330,6 +2336,245 @@ mod benchmarks {
 		assert_eq!(meter.consumed(), <T as Config>::WeightInfo::v1_migration_step() * 2);
 	}
 
+	// Helper function to create a test signer for finalize_block benchmark
```
Done

pgherveou commented on substrate/frame/revive/src/lib.rs:
```diff
@@ -623,8 +625,17 @@ pub mod pallet {
 		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+			// Account for the fixed part of the costs incurred in `on_finalize`.
+			frame_system::Pallet::<T>::register_extra_weight_unchecked(
+				<T as Config>::WeightInfo::on_finalize_block_fixed(),
```
not sure what  register_extra_weight_unchecked is doing but we should just return the Weight here.

We should account for killing the storage, but after our last sync discussion,  not sure we need all these kill.  

All these storage will be either cleared up or overwrite on_finalize

pgherveou commented on substrate/frame/revive/src/lib.rs:
```diff
@@ -1454,16 +1471,23 @@ where
-						// payload. Instead, use an empty vector. The signed transaction
-						// will be provided by the user when the tx is submitted.
-						payload: Vec::new(),
+						// Payload is needed to determine costs.
+						// Safe to unwrap(), it is asserted when converting to TransactionUnsigned
+						// above
+						payload: unsigned_tx.clone().unwrap().dummy_signed_payload(),
```
not a big fan of these unwrap, can't we just compute that unsigned_payload once and not deal with optional?

pgherveou commented on substrate/frame/revive/src/lib.rs:
```diff
@@ -1799,6 +1829,13 @@ impl<T: Config> Pallet<T> {
 		Some(T::AddressMapper::to_address(&account_id))
 	}
 
+	#[cfg(feature = "runtime-benchmarks")]
```
let's just update the function to return an H160 and default_or_unwrap
we will keep the block_author function in runtime_api as it is for now to not create any breaking change in the substrate rpc APIS
